src/routes/auth/auth.service.ts

```diff
@@ -0,0 +1,156 @@
+   *
+   * The following adhere to JWT standard {@link https://datatracker.ietf.org/doc/html/rfc7519}
+   *
+   * @returns accessToken - JWT access token
+   * @returns tokenType - token type ('Bearer') to be used in the `Authorization` header
+   * @returns notBefore - expoch from when token is valid (if applicable, otherwise null)
+   * @returns expiresIn - time in seconds until the token expires (if applicable, otherwise null)
+   */
+  async verify(args: {
+    request: Request;
+    verifyAuthMessageDto: VerifyAuthMessageDto;
+  }): Promise<{
+    accessToken: string;
+    tokenType: string;
+    notBefore: number | null;
+    expiresIn: number | null;
+  }> {
+    const isAuthorized = await this.isAuthorized(
+      args.verifyAuthMessageDto,
+    ).catch(() => false);
+
+    if (!isAuthorized) {
+      throw new UnauthorizedException();
+    }
+
+    const { message } = args.verifyAuthMessageDto;
+
+    const dateWhenTokenIsValid = message.notBefore
+      ? new Date(message.notBefore)
+      : null;
+    const dateWhenTokenExpires = message.expirationTime
+      ? new Date(message.expirationTime)
+      : null;
+
+    const secondsUntilTokenIsValid = dateWhenTokenIsValid
+      ? this.getSecondsUntil(dateWhenTokenIsValid)
+      : null;
+    const secondsUntilTokenExpires = dateWhenTokenExpires
+      ? this.getSecondsUntil(dateWhenTokenExpires)
+      : null;
+
+    const accessToken = this.jwtService.sign(message, {
+      ...(secondsUntilTokenIsValid !== null && {
+        notBefore: secondsUntilTokenIsValid,
+      }),
+      ...(secondsUntilTokenExpires !== null && {
+        expiresIn: secondsUntilTokenExpires,
+      }),
+    });
+
+    return {
+      tokenType: AuthService.AUTH_TOKEN_TOKEN_TYPE,
+      accessToken,
+      notBefore: dateWhenTokenIsValid?.getTime() ?? null,
+      expiresIn: secondsUntilTokenExpires,
+    };
+  }
+
+  /**
+   * Verifies that a message is valid according to its expiration date,
+   * signature and nonce.
+   *
+   * @param args.message - SiWe message in object form
+   * @param args.signature - signature from signing the message
+   *
+   * @returns boolean - whether the message is valid
+   */
+  private async isAuthorized(args: {
+    message: SiweMessage;
+    signature: `0x${string}`;
+  }): Promise<boolean> {
+    const isExpired =
+      !!args.message.expirationTime &&
+      new Date(args.message.expirationTime) < new Date();
+
+    const isValidSignature = await this.authRepository
+      .verifyMessage(args)
+      // Don't prevent nonce from being deleted
+      .catch(() => false);
+
+    const cacheDir = CacheRouter.getAuthNonceCacheDir(args.message.nonce);
+    const cachedNonce = await this.cacheService.get(cacheDir);
+    const isValidNonce = cachedNonce === args.message.nonce;
+
+    // Delete nonce from cache to prevent replay attacks
+    await this.cacheService.deleteByKey(cacheDir.key);
```

Comment:
This breaks the abstraction that we have in the routes when it comes to accessing the Datasources. This should go through the domain layer instead.

Reply:
Moved to `AuthRepository` in 52e1c66649c3a40858d05b1258748cfe621b3b26.